test/unit/transaction_test.js

```diff
@@ -634,6 +635,85 @@ describe('Transaction', function() {
       expect(() => tx.getClaimableBalanceId(-1)).to.throw(/index/);
     });
   });
+
+  describe('preconditions', function() {
+    const address = 'GA7QYNF7SOWQ3GLR2BGMZEHXAVIRZA4KVWLTJJFC7MGXUA74P7UJVSGZ';
+
+    const source = new StellarBase.Account(address, '1234');
+    const makeBuilder = function() {
+      return new StellarBase.TransactionBuilder(source, {
+        fee: StellarBase.BASE_FEE,
+        networkPassphrase: StellarBase.Networks.TESTNET,
+        withMuxing: true
+      });
+    };
+
+    describe('timebounds', function() {
+      it('Date', function() {
+        let now = new Date();
+        let tx = makeBuilder()
+          .setTimebounds(now, now)
+          .build();
+        expect(tx.timeBounds.minTime).to.eql(
+          `${Math.floor(now.valueOf() / 1000)}`
+        );
+        expect(tx.timeBounds.maxTime).to.eql(
+          `${Math.floor(now.valueOf() / 1000)}`
+        );
+      });
+
+      it('number', function() {
+        let tx = makeBuilder()
+          .setTimebounds(5, 10)
+          .build();
+        expect(tx.timeBounds.minTime).to.eql('5');
+        expect(tx.timeBounds.maxTime).to.eql('10');
+      });
+    });
+
+    it('ledgerbounds', function() {
+      let tx = makeBuilder()
+        .setTimeout(5)
+        .setLedgerbounds(5, 10)
+        .build();
+
+      expect(tx.ledgerBounds.minLedger).to.equal(5);
+      expect(tx.ledgerBounds.maxLedger).to.equal(10);
+    });
+
+    it('minAccountSequence', function() {
+      let tx = makeBuilder()
+        .setTimeout(5)
+        .setMinAccountSequence('5')
+        .build();
+      expect(tx.minAccountSequence).to.eql('5');
+    });
+
+    it('minAccountSequenceAge', function() {
+      let tx = makeBuilder()
+        .setTimeout(5)
+        .setMinAccountSequenceAge(5)
+        .build();
+
+      expect(tx.minAccountSequenceAge.toString()).to.equal('5');
+    });
+
+    it('minAccountSequenceLedgerGap', function() {
+      let tx = makeBuilder()
+        .setTimeout(5)
+        .setMinAccountSequenceLedgerGap(5)
+        .build();
+      expect(tx.minAccountSequenceLedgerGap).to.equal(5);
+    });
+
+    it('extraSigners', function() {
+      let tx = makeBuilder()
+        .setTimeout(5)
+        .setExtraSigners([address])
```

Comment:
A quick question, maybe not related to this PR. Here, `address === 'GA7QYNF7SOWQ3GLR2BGMZEHXAVIRZA4KVWLTJJFC7MGXUA74P7UJVSGZ'`, but it should actually be an address starting with `P`, right? We currently have four types of SignerKey: ED25519, PRE_AUTH_TX, HASH_X and ED25519_SIGNED_PAYLOAD, can the SignerKey used here only be of type ED25519_SIGNED_PAYLOAD?

Reply:
All signers can be used here, but yeah, as Paul said, we’re keeping it simple for now and using the XDR directly, but I’ll revisit this to create a simple abstraction layer to it from the strkey addresses.